hacks/prune-rhcos-snapshots/prune-rhcos-snapshots.py

```diff
@@ -0,0 +1,255 @@
+                            installer_ami_ids[region_name]['images'] = []
+                        ami_id = region_entry['image']
+                        installer_ami_ids[region_name]['images'].append(ami_id)
+                        must_preserve.add(f'{region_name}:{ami_id}')
+
+            rhcos_in_use[payload_key] = {
+                'payload_version': machine_os_rhcos_version,
+                'installer_amis': installer_ami_ids
+            }
+
+            with riu_path.open('a+') as f:
+                cache_entry = yaml.dump({
+                    payload_key: rhcos_in_use[payload_key]
+                })
+                # Write, line by line, to the cache file
+                f.write(f'{cache_entry}\n')
+
+    client = boto3.client('ec2')
+    all_aws_regions = [region['RegionName'] for region in client.describe_regions()['Regions']]
+
+    ami_analysis = dict()
+    for aws_region in all_aws_regions:
+        region_ami_analysis = dict()
+        ami_analysis[aws_region] = region_ami_analysis
+
+        # Ensure there are recycle bin rules setup to preserve AMI/snapshots
+        # rbin_client = boto3.client('rbin', region_name=aws_region)  # Recycle bin client
+        # snapshot_rules = rbin_client.list_rules(ResourceType='EBS_SNAPSHOT')
+        # if len(snapshot_rules['Rules']) == 0:
+        #     # Create snapshot rule for X day retention
+        #     create_recycle_bin_rule(rbin_client, 'EBS_SNAPSHOT')  # requires new privs
+        #
+        # image_rules = rbin_client.list_rules(ResourceType='EC2_IMAGE')
+        # if len(image_rules['Rules']) == 0:
+        #     # Create ec2 image rule for X day retention
+        #     create_recycle_bin_rule(rbin_client, 'EC2_IMAGE')  # requires new privs
+
+        region_client = boto3.client('ec2', region_name=aws_region)
+        ec_resource = boto3.resource('ec2', region_name=aws_region)
+        print(f'Querying for AMIs in {aws_region}..')
+        images = region_client.describe_images(Owners=['self']).get('Images')
+
+        for image in images:
+            image_id = image['ImageId']
+            image_name = image.get('Name', None)
+            image_description = image.get('Description', None)
+            image_creation = image['CreationDate']
+            image_tags = image.get('Tags', [])
+
+            print(f'  Checking AMI {image_id} ({image_name})')
+            # Build up a string we are certain will contain the RHCOS version
+            ami_info = f'{image_name} {image_description} {image_tags}'
+            matches_payload_keys = list()
+            for payload_key, payload_key_references in rhcos_in_use.items():
+                if payload_key_references['payload_version'] in ami_info:
+                    print(f'Will not clean up RHCOS AMI {image_id} ({image_name}) since it is referenced by a release controller release payload {payload_key}')
+                    matches_payload_keys.append(payload_key)
```

Comment:
do we actually care about release payload references to AMIs? we only use AMIs at install time, right? and i'm pretty sure we don't reference AMIs anywhere besides build metadata and the installer so don't we only need to keep things that the installer refers to?

the difference isn't going to be huge though if you want to be cautious here and keep both.

Reply:
> the difference isn't going to be huge though if you want to be cautious here and keep both.
I don't think we do need them. So, yes, this is just a cautionary step.